### ISSUE TITLE
Interrupting animations with new animations

### ISSUE THREAD

**Issue by edemaine** [wout/svg.js#402]
I have several apps where I call .animate() to change any visual features of objects.  These changes are triggered by mouse over, mouse exit, click, etc. events.  Unfortunately, I find that if I create a new animation with .animate(), and another animation on the same object is still running, the old animation gets interrupted, so the older visual changes never get implemented, leaving to an inconsistent visual state.

1. I'm not sure this is the intended behavior.  Other behaviors I could see: automatically jump any existing animation to the end (fast forward), or start a new animation to run in parallel to the old one.

2. Even if it is intended behavior, the natural way to get the fast-forward behavior (from the documentation) seems to be to call x.stop(true) before creating a new x.animate().  However, x.stop(true) just calls x.animate(0) which also doesn't seem to fast forward the animation (despite the documentation).

3. The fix I found is to call x.at(1) before x.stop().  Note that .at is not in the documentation.

**Comment by Fuzzyma** [comment]
This is a thing for svg.js v3 which I try to solve till then.
My recommendation would be to alter the animation module to the following:

- animate calls which change the same attribute become overwritten (without interuption)
- you can mix up and time multiple animates without anything struggling
- stop() stoppes the animation right at the point where it is
- finish() moves the animation to the end (same as at(1))
- resume() will unpause the animation (pause = stop)
- reverse() plays the animation in reverse from the point where it currently is
- times() speeds up or slows down the animation by applying a factor to the position value

Feel free to add your own thoughts about this.
I did not think that much about it till now but I seriously plan to implement this

**Comment by edemaine** [comment]
Sounds great, especially that second bullet!

#225 is possibly related.  Actually, .after gets somewhat ambiguous once you have multiple animations with different completion times: does it mean after all animations, or the "current" one?  Consider:

```
x.animate(1000).fill('black')
x.animate(500).stroke('blue')
x.after(function(){ x.animate(500).opacity(0.5) })
```

I guess the .after applies to the current animation, so happens after 500ms?

In some ways I would find it cleaner if .animate() returns an animation object which is not the original SVG object.  This would make it possible to pause, cancel, etc. individual animations and not just the whole thing.  You could still have stop() etc. on the SVG object, which applies to all animations on that object.  Hmm, I suppose you could add .after() to both as well, to be able to specify either event.

This would let you do some neat things like this: (in response to #225)

```
function chain (anims) {
  for (var i = 0; i < anims.length; i++) {
    anims[i].stop()
    anims[i].after(anims[i+1].resume)
  }
  anims[0].resume()
}
chain ([
  x.animate().fill('black'),
  x.animate().stroke('purple'),
  y.animate().opacity(0.5)
])
```

**Comment by Fuzzyma** [comment]
animate() does return an instance of `fx` and not of the element itself so you actually can do your stuff on that fx-object which is possible already. That also explains when `after()` is called (after this specific animation).

The code you specied is also not that elegant. 
I would prefer the `enque()` method which adds a new animation to the animation chain.
However I also would like it if every call to animation is chained instead of called directly (somewhat like jquery does). Many stuff to think about

**Comment by edemaine** [comment]
Oh, cool.  Sorry I didn't realize the ```fx``` objects.

I wonder if we could take some inspiration from PowerPoint, which has extensive animation features.  In it, every animation has a choice of "animate on click" (not relevant here), "animate after previous", and "animate with previous" (simultaneous).  And there is an order to all animations.  So we could imagine a boolean argument to .animate() to specify simultaneous with previous animations or enqueued after previous animations -- or separate .animate() and .animateafter() methods.  Both are appending to the order of animations, but marked either simultaneous or waiting.  You could alternatively call them .animate() and .enqueue() of course, viewing it as a queue of simultaneous animations...  In either case, in advanced uses, you could manipulate the animation queue, I guess like ```svgobject.fx``` currently gives you access to the sole animation.

PowerPoint also has a "delay" before the actual animation start, which lets you e.g. start one animation in the middle of another.  This seems like it would be a useful optional argument to .animate() (and .enqueue()).  For example:

```
x.animate(1000).stroke('purple')
x.animate(500,delay=250).fill('black')
```
where the latter line would act something like
```x.at(0.25, function(){ x.animate(500).fill('black') })``` or
```setTimeout(function(){ x.animate(500).fill('black') }, 250)```.

**Comment by Fuzzyma** [comment]
You should consider reading the docs ;)
delay is already implemented.

I think when you comapre PowerPoint to svgjs its more like with the Apple and Oranges.
There is not rly a "start with the previos" because it will start whenever called which can be immediately or some ms after. However the general idea is the same.
I see some problems when you have a very mixed timeline:

```
animation1start..................animation1end -> animation2start.......................animation2end
                          animation3start.......................................animation3end
```

You cannot rly have an order (or queue) for this case. Maybe it has to be possible to create a new queue for every fx instance. The relation then would be "animation1 is followed by animation2" and "animation3 is followed by animation4". So it's more a relation between the animations instead of a simple queue where everything is linear and in order

**Comment by edemaine** [comment]
Wow, sorry.  I've read through the docs several times lately, but I clearly missed the delay argument.  Cool feature! :-)

Yes, there are many complicated scenarios, though it may make sense to focus on the "common" ones, whichever those are.  Perhaps parallel and sequential animation, either wanting to specify them upfront as a single combined animation, or additional components unexpectedly/last-minute as in my application?  I can also imagine it being common to want to animate one object x and sequentially follow that by an animation of another object y.  So it doesn't necessarily make sense to have the animation chain tied to a specific SVG object, but rather be able to connect between them...  (I get the feeling we are re-inventing .after() though...)

Here's a different/related idea.  What if there were an fx group object (similar to SVG groups), which could contain multiple fx's?  This could make it easy to, say, ```x.animate(500).fill('black')``` in parallel with ```y.animate(600).stroke('purple')``` and then schedule an event that happens after both of them.  (Create an .after trigger on the group.)  Maybe this concept could be useful for other things we're trying to do...

By the way, I haven't tried it yet, but I imagine sequential .after() chains would be especially clean in IcedCoffeeScript.  I think it would be like this:

```
await x.animate(500).fill('black').after defer()
await y.animate(500).stroke('purple').after defer()
...whatever should happen after those two animations...
```

**Comment by Fuzzyma** [comment]
Look at the 3.0.0 branch for progress.

I dont know IcedCoffeeScript - especially i am not aware of the `await` method.
There is no way to mirror this behavior in native js.
However I like the `defer` thing. Any Idea on how to implement that?

**Comment by edemaine** [comment]
Wow, 3.0.0's fxnew.js looks very cool -- powerful chaining of effects

I'm a little confused whether `this._next` is supposed to be a last pointer or something else.  The `last` method suggests that `this._next` is not a last pointer, but `enqueue` uses `next` instead of `last` so I'm confused...

My IcedCoffeeScript code above is equivalent to this in plain JavaScript:

```
x.animate(500).fill('black').after(function(){
  y.animate(500).stroke('purple').after(function(){
    ...whatever should happen after those two animations...
  })
})
```

What's cool about IcedCoffeeScript is you can do for loops and such over await/defer pairs, which is particularly annoying in regular JavaScript.

Anyway, it seems like in the 3.0.0 branch you can do something like this?

```
x.animate(500).fill('black').enqueue(
  y.animate(500).stroke('purple')  // does this need to be .stop()'d?
).afterAll(function() {
  ...whatever should happen after those two animations...
})
```

which looks pretty nice!  And it's even cleaner if you're manipulating the same object:

```
x.animate(500).fill('black').enqueue(500).stroke('purple').afterAll(function() { ... })
```

Did I get it right?  I haven't actually tried using the code, just reading it.

**Comment by Fuzzyma** [comment]
I have to disappoint you about 
> ```javascript
> x.animate(500).fill('black').after(function(){
>   y.animate(500).stroke('purple').after(function(){
>     ...whatever should happen after those two animations...
>   })
> })
> ```

Its the other syntax you used in the last code block.
However when you replace `enqueue` with `after` it would maybe work.
The enqueue function is using next because you are already manipulating the last element of the queue. So the next-pointer which is null at this place is filled with a new animation which is set in the enqueue method.
One thing to mention is, that you are never manipulating the same object. You always get back an instance of `fx` but when you call `enqueue()` a new instance is added to the queue which is retuned and on which you are working then.

Ah - and you cant animate "color-words" because svg.js has no map for this. Use hex or rgb instead. Also: You can only animate attributes which are formerly set (its like css transitions)

**Comment by edemaine** [comment]
Right, the code using `.after` was written in the old (v2) style.  Do the `.enqueue` style examples I sketched work, other than using named colors (oops!)?

**Comment by Fuzzyma** [comment]
Implemented in 2.3.0
